### PR TITLE
[CWS] Converting SecAgent Start command to fx

### PR DESCRIPTION
**Pull request by modernplumbing**
### What does this PR do?

Moved the SecAgent start command to fx.

**Questions for Components team unless otherwise marked**
1. Log Flushing at the root command level before https://datadoghq.atlassian.net/browse/ASC-340 gets merged. Right now, we’re flushing the log as a persistent [post run on the root command](https://github.com/DataDog/datadog-agent/blob/194c888096366eded34905e03f59cf31814b8239/cmd/security-agent/command/command.go#L53). However, I believe that’s a different logger than the one being supplied through fx like [here](https://github.com/DataDog/datadog-agent/blob/46fde246fd67d950d4bc09dc8d269ce57594a44b/cmd/security-agent/subcommands/runtime/command.go#L97). I could flush at the end of every one shot fnc, but I’m wondering if there’s a way I can do it at the root command to not duplicate code.
- Answer: Doing it manually for now.

2. Another question I have is that I have a [package](https://github.com/DataDog/datadog-agent/blob/46fde246fd67d950d4bc09dc8d269ce57594a44b/cmd/security-agent/command/logs_context.go) that does “logging” and “configuring” with methods that are not part of the component interfaces. That package is called [here](https://github.com/DataDog/datadog-agent/blob/efeab9a2f591cf5b95c1a4438d9abe032cf9e5fa/cmd/security-agent/subcommands/compliance/compliance.go#L32). So when I use the pkg/util/log and config packages outside of the instantiated component, am I working with separate instances of the logger and config? In the case of config, is everything eventually part of the config.Datadog global object?
- Another [example](https://github.com/DataDog/datadog-agent/blob/194c888096366eded34905e03f59cf31814b8239/cmd/security-agent/subcommands/start/command.go#L182):
```
	port := config.GetString("security_agent.expvar_port")
	pkgconfig.Datadog.Set("expvar_port", port)
```
 - Answer: pkconfig.Datadog is equivalent to the config component

3. How bad is it that we seem to have multiple instances of a logger floating around? See [this](https://github.com/DataDog/datadog-agent/blob/celia.yuen/start-cmd-fx/cmd/security-agent/config/config.go#L26) as an example of a non component logger.

4. [For SecAgent team] In the new Merge Config process, the output of some commands are now the following when an expected config file is missing. Is this UX ok?
```
vagrant@agent-dev-ubuntu-22:/git/datadog-agent$ sudo bin/security-agent/security-agent status
Error: unable to load Datadog config file: open /etc/datadog-agent/datadog.yaml: no such file or directory
vagrant@agent-dev-ubuntu-22:/git/datadog-agent$ sudo bin/security-agent/security-agent version
Error: unable to load Datadog config file: open /etc/datadog-agent/datadog.yaml: no such file or directory
```
- Answer: yes

#### Additional changes
Part of this change included changes to how we set up the SecAgent config.

**Config loading flow:**

The default config paths are path.Join(common.DefaultConfPath, "datadog.yaml") and path.Join(common.DefaultConfPath, "security-agent.yaml").

**Original**

Cases
1.  Default: Load datadog.yaml, then merge security-agent.yaml.   
        - Successfully run agent
7. Want default, but either datadog.yaml or security-agent.yaml are inaccessible
	- Continue to run agent with no warning. Technically a success 
8.  No config files (only possible if user-defined, i.e. `cfgpath=`)
        - Exit
9.  User defined paths to override the defaults
        - Success
10.  User defined paths, but at least one is inaccessible
       - fmt.Printf("Warning: unable to access %s\n", configurationFilename)
       - If at least one config is accessible, continue to run agent. Technically a success.

**New**
1.  Default: Load datadog.yaml, then merge security-agent.yaml.   
        - Successfully run agent
5. Want default, but ~~either datadog.yaml or~~ security-agent.yaml ~~are~~ is inaccessible
	- Continue to run agent with no warning. Technically a success
	- **CHANGE**: If datadog.yaml is missing, because it's the first of the defaults in the slice of defaults, immediately exit
6.  No config files (only possible if user-defined, i.e. `cfgpath=`)
        - Exit
7.  User defined paths to override the defaults
        - Success
8.  User defined paths, but at least one is inaccessible
      - **CHANGE**: If the first of the passed in config paths are inaccessible, immediately exit

Config subcommand here: https://github.com/DataDog/datadog-agent/pull/15153/files

### Additional Notes

Still need to convert the config subcommand to fx. 

### Describe how to test/QA your changes
https://datadoghq.atlassian.net/wiki/spaces/CWS/pages/2772631553/CLI+QA+Checklist

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by modernplumbing** [comment]
Current dependencies on/for this PR:
* main
  * **PR #14833** <a href="https://app.graphite.dev/github/pr/DataDog/datadog-agent/14833" target="_blank">Graphite</a> 
    * **PR #14814** <a href="https://app.graphite.dev/github/pr/DataDog/datadog-agent/14814" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/DataDog/datadog-agent/14814?utm_source=stack-comment).

File: cmd/security-agent/app/app.go [hunk 1]
```diff
@@ -259,7 +260,9 @@ func RunAgent(ctx context.Context, pidfilePath string) (err error) {
 		}
 	}
 
-	complianceAgent, err := compliance.StartCompliance(hostnameDetected, stopper, statsdClient)
+	var log complog.Component
+	var config compconfig.Component
+	complianceAgent, err := compliance.StartCompliance(log, config, hostnameDetected, stopper, statsdClient)
```
Comment (ogaca-dd) [inline cmd/security-agent/app/app.go:259]: TODO

File: cmd/cluster-agent/subcommands/compliance/command.go [hunk 2]
```diff
@@ -32,7 +32,8 @@ func Commands(globalParams *command.GlobalParams) []*cobra.Command {
 	}
 
 	// TODO: The SecAgent Check package should be a component
-	complianceCmd.AddCommand(check.CommandsWrapped(bundleParams)...)
+	// The globalParams input should always be nil here. This is used to differentiate between entering from Cluster-Agent or the Security-Agent
```
Comment (ogaca-dd) [inline cmd/cluster-agent/subcommands/compliance/command.go:35]: Passing `nil` to tell if the function is called from the Cluster Agent is not explicit from the point of view of the implementation of `CommandsWrapped`.
What is the expected differences of behaviour for  `CommandsWrapped` when calling from the Cluster or the from Security Agent?
My understanding is that `CommandsWrapped` should override the logger and the configuration for the cluster Agent.

Comment (modernplumbing) [reply cmd/cluster-agent/subcommands/compliance/command.go]: Yea, even if this is pending refactoring, I should make it more explicit. I added a flag to determine if it's coming from cluster-agent or security-agent. CommandsWrapped uses either the params from the security-agent to instantiate the logger and config, or the whole bundle param from cluster-agent.

File: cmd/security-agent/command/logs_context.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+func NewLogContextCompliance() (*logsconfig.Endpoints, *client.DestinationsContext, error) {
+	logsConfigComplianceKeys := logsconfig.NewLogsConfigKeys("compliance_config.endpoints.", pkgconfig.Datadog)
+	return NewLogContext(logsConfigComplianceKeys, "cspm-intake.", "compliance", logsconfig.DefaultIntakeOrigin, logs.AgentJSONIntakeProtocol)
+}
+
+// This function will only be used on Linux. The only platforms where the runtime agent runs
+func NewLogContextRuntime() (*logsconfig.Endpoints, *client.DestinationsContext, error) {
+	logsConfigComplianceKeys := logsconfig.NewLogsConfigKeys("runtime_security_config.endpoints.", pkgconfig.Datadog)
+	return NewLogContext(logsConfigComplianceKeys, "runtime-security-http-intake.logs.", "logs", cwsIntakeOrigin, logsconfig.DefaultIntakeProtocol)
```
Comment (lebauce) [inline cmd/security-agent/command/logs_context.go:29]: nit: rename to logsRuntimeConfigKeys

Comment (modernplumbing) [reply cmd/security-agent/command/logs_context.go:29]: done

File: releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+---
+enhancements:
+  - |
+    The security agent will fail quickly if the first of either the default, or if they exist, user-provided override config files are inaccessible.
```
Comment (maycmlee) [inline releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml:4]:
```suggestion
    The security agent will fail immediately if the first of either the default, or if they exist, user-provided override config files are inaccessible.
```

Comment (maycmlee) [reply releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml:4]: I'm having a hard time understanding this:
"if the first of either the default, or if they exist, user-provided override config files are inaccessible."

File: releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml [hunk 5]
```diff
@@ -0,0 +1,6 @@
...
+    The default config files are still datadog.yaml and security-agent.yaml, in that order.
```
Comment (maycmlee) [inline releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml:5]:
```suggestion
    The default config files are still datadog.yaml and security-agent.yaml, and loaded in that order.
```

File: releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+    Previously, if the security agent was started with at least one accessible config file, no matter the order, the agent would still be able to start.
```
Comment (maycmlee) [inline releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml:6]:
```suggestion
    Previously, if the security agent was started with at least one accessible config file, no matter what order the files were loaded, the Datadog Agent would still be able to start.
```

Comment (maycmlee) [reply releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml:6]: I'm assuming the agent mentioned here is the Datadog Agent, let me know if that's wrong.

**Review by maycmlee — COMMENTED** [review]
Left some feedback and questions from docs.

File: releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml (hunk 6 above)
Comment (modernplumbing) [reply releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml:6]: It's the security agent. The new sentence is 
```
    Previously, if the security agent was started with at least one accessible config file, no matter what order, the security agent would still be able to start.
```

File: releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml (hunk 4 above)
Comment (modernplumbing) [reply releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml:4]: Is this clearer?
```
The security agent will fail immediately if the first of either the default config files or user-provided override config files are inaccessible.

```

File: cmd/cluster-agent/subcommands/compliance/command.go (hunk 2 above)
Comment (ogaca-dd) [reply cmd/cluster-agent/subcommands/compliance/command.go]: You can use a lambda function to configure `BundleParams`.
Note:  The values inside `globalParams` are not yet set when calling `SecurityAgentCommands` but can safely used inside `RunE`. 
```go
func SecurityAgentCommands(globalParams *command.GlobalParams) []*cobra.Command {
	return commandsWrapped(func() core.BundleParams {
		return core.BundleParams{
			ConfigParams: config.NewSecurityAgentParams(globalParams.ConfigFilePaths),
			LogParams:    log.LogForOneShot(command.LoggerName, "info", true),
		}
	})
}

func ClusterAgentCommands(bundleParams core.BundleParams) []*cobra.Command {
	return commandsWrapped(func() core.BundleParams {
		return bundleParams
	})
}

func commandsWrapped(bundleParamsFactory func() core.BundleParams) []*cobra.Command {
	checkArgs := &CliParams{}

	cmd := &cobra.Command{
		Use:   "check",
		Short: "Run compliance check(s)",
		Long:  ``,
		RunE: func(cmd *cobra.Command, args []string) error {
			checkArgs.args = args
			bundleParams := bundleParamsFactory()

			if checkArgs.verbose {
				bundleParams.LogParams = log.LogForOneShot(bundleParams.LogParams.LoggerName(), "trace", true)
			}
  / ...
}
```

**Review by ogaca-dd — APPROVED** [review]
LGTM

File: releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml (hunk 4 above)
Comment (maycmlee) [reply releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml]: Would taking out "the first of" affect the meaning? That part is kind of confusing. So:

```
The security agent will fail immediately if either the default config files or user-provided override config files are inaccessible.
```

Comment (modernplumbing) [reply releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml]: The the default config files are 2 files, and the user provided files are 0+. The security agent fails if the first of either of these sets are inaccessible.

File: releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml (hunk 6 above)
Comment (maycmlee) [reply releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml]: That sentence works, just added a small edit, so it's "no matter in what order". I updated my suggested change.

Comment (modernplumbing) [reply releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml]: implementing!

File: releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml (hunk 4 above)
Comment (maycmlee) [reply releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml]: Thanks, that's really helpful. I think using some of that will make it clearer for the user. Does this make sense and is it correct? 

> There are two default config files. Optionally, you can also provide override config files. The security agent fails if the first config file of either of these sets are inaccessible. In the default case, there is the datadog.yaml and security-agent.yaml, and in that order. If datadog.yaml is inaccessible, the security agent fails immediately. If you provide overrides, like foo.yaml and bar.yaml, the security agent only fails immediately if foo.yaml is inaccessible.

Comment (modernplumbing) [reply releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml]: how about this? i wanted to highlight the change

```
There are two default config files. Optionally, you can provide override config files. 

The change in this release is that the security agent now fails if the first config file of either of these sets are inaccessible. In the default case, the config files are datadog.yaml and security-agent.yaml, and in that order. If datadog.yaml is inaccessible, the security agent fails immediately. If you provide overrides, like foo.yaml and bar.yaml, the security agent fails immediately if foo.yaml is inaccessible. 

In both sets, if any following config files are missing, the security agent continues to attempt to start up, with a log message about an inaccessible config file. 

Previously, the security-agent would continue attempting to start up if the first config file was inaccessible.  
```

Comment (maycmlee) [reply releasenotes/notes/runtime-security-config-file-handling-a9508fda817b9b30.yaml]: That looks good!

**Review by paulcacheux — APPROVED** [review]

**Review by maycmlee — APPROVED** [review]
👍 for docs

**Review by vboulineau — APPROVED** [review]

**Pull request merged**